simplecpp.cpp

```diff
@@ -1716,42 +1716,127 @@ namespace simplecpp {
+        else
+            return NAME;
+    }
+    typedef std::vector<std::string> PathPartsT; // vector is not really a must-have...
 
-        // remove "./"
-        pos = 0;
-        while ((pos = path.find("./",pos)) != std::string::npos) {
-            if (pos == 0 || path[pos - 1U] == '/')
-                path.erase(pos,2);
-            else
-                pos += 2;
+    /// split path string into components - neglecting the path separators!
+    static PathPartsT SplitPath(const std::string& path)
+    {
+        PathPartsT pathParts;
+        std::string subPath;
+        for (std::string::const_iterator it=path.begin(); it!=path.end(); ++it) {
+            if (*it==SEP) {
+                if (!subPath.empty()) {
+                    pathParts.push_back(subPath);
+                    subPath.clear();
+                }
+            } else
+                subPath+=*it;
         }
+        if (!subPath.empty())
+            pathParts.push_back(subPath);
 
-        // remove "xyz/../"
-        pos = 1U;
-        while ((pos = path.find("/../", pos)) != std::string::npos) {
-            const std::string::size_type pos1 = path.rfind('/', pos - 1U);
-            if (pos1 == std::string::npos) {
-                path.erase(0,pos+4);
-                pos = 0;
-            } else {
-                path.erase(pos1,pos-pos1+3);
-                pos = std::min((std::string::size_type)1, pos1);
+        return pathParts;
+    }
+
+    // Normalize path: resolve .., remove meaningless .
+    static void NormalizePath(PathPartsT& pathParts)
+    {
+        PathPartsT::iterator current=pathParts.begin();
+        while (current!=pathParts.end()) {
+            switch (DeterminePathComponentType(*current)) {
+            case CURDIR:
+                current=pathParts.erase(current);
+                break;
+            case UPDIR:
+                if (current==pathParts.begin()) {
+                    ++current;
+                } else {
+                    switch (DeterminePathComponentType(*(current-1))) {
+                    case UPDIR:
+                        ++current;
+                        break;
+                    case CURDIR:
+                    case NAME:
+                        current=pathParts.erase(current-1, current+1);
+                        break;
+                    }
+                }
+                break;
+            case NAME:
+                ++current;
+                break;
             }
         }
+    }
 
-        return realFilename(path);
+    std::string simplifyPath(std::string path)
+    {
+        if (path.empty())
+            return "";
+        // detect UNC path on DOS/Windows
+        const bool isUnc = (0==path.compare(0,2,"\\\\")) || (0==path.compare(0,2,"//"));
+        // replace backslash separators
+        std::replace(path.begin(), path.end(), '\\', SEP);
+        // detect: DOS/Windows path including drive letter
+        const bool isAbsolute=path[0]==SEP;
```

Comment:
I think `isAbsolute` could be `(path[0]==SEP || DOSDriveLetterPathAbsolute(path))` without changing the meaning.